src/java.base/share/classes/java/lang/reflect/Proxy.java

```diff
@@ -1120,4 +1139,199 @@ public static InvocationHandler getInvocationHandler(Object proxy)
+        Class<?> declaringClass = method.getDeclaringClass();
+        if (!declaringClass.isInterface()) {
+            throw new IllegalArgumentException("\"" + method +
+                    "\" is not a method declared in the proxy class");
+        }
+
+        List<Class<?>> proxyInterfaces = Arrays.asList(proxyClass.getInterfaces());
+        // the method's declaring class is a proxy interface
+        if (proxyInterfaces.contains(declaringClass))
+            return declaringClass;
+
+        Deque<Class<?>> deque = new ArrayDeque<>();
+        Set<Class<?>> visited = new HashSet<>();
+        boolean indirectMethodRef = false;
+        for (Class<?> intf : proxyInterfaces) {
+            assert intf != declaringClass;
+            visited.add(intf);
+            deque.add(intf);
+
+            Class<?> c;
+            while ((c = deque.poll()) != null) {
+                if (c == declaringClass) {
+                    try {
+                        // check if this method is the resolved method if referenced from
+                        // this proxy interface (i.e. this method is not implemented
+                        // by any other superinterface)
+                        Method m = intf.getMethod(method.getName(), method.getParameterTypes());
+                        if (m.getDeclaringClass() == declaringClass) {
+                            return intf;
+                        }
+                        indirectMethodRef = true;
+                    } catch (NoSuchMethodException e) {}
+
+                    // skip traversing its superinterfaces
+                    // another proxy interface may extend it and so
+                    // the method's declaring class is left unvisited.
+                    continue;
+                }
+                // visit all superinteraces of one proxy interface to find if
+                // this proxy interface inherits the method directly or indirectly
+                visited.add(c);
+                for (Class<?> superIntf : c.getInterfaces()) {
+                    if (!visited.contains(superIntf) && !deque.contains(superIntf)) {
+                        if (superIntf == declaringClass) {
+                            deque.addFirst(superIntf);
+                        } else {
+                            deque.add(superIntf);
+                        }
+                    }
+                }
+            }
+        }
+
+        throw new IllegalArgumentException("\"" + method + (indirectMethodRef
+                ? "\" is overridden directly or indirectly by the proxy interfaces"
+                : "\" is not a method declared in the proxy class"));
+    }
+
+    /**
+     * Returns a Lookup object for the lookup class which is the class of this
+     * proxy instance.
+     *
+     * @return a lookup for proxy class of this proxy instance
+     */
+    private static MethodHandles.Lookup proxyClassLookup(MethodHandles.Lookup caller, Class<?> proxyClass) {
```

Comment:
The method description could be a bit clearer. It invokes the proxy's proxyClassLookup method to get a Lookup on the proxy class ("this proxy instance" is just a bit confusing as proxyClassLookup is static).
I guess the caller parameter isn't really needed as it could be crated in proxyClassLookup.

Reply:
The caller parameter is just another level of defense.   I updated as:

```
     /**
-     * Returns a Lookup object for the lookup class which is the class of this
-     * proxy instance.
+     * This method invokes the proxy's proxyClassLookup method to get a
+     * Lookup on the proxy class.
      *
      * @return a lookup for proxy class of this proxy instance
      */
```

Reply:
Much better, thanks!